deviations/viewbase.py

```diff
@@ -0,0 +1,295 @@
+from itertools import groupby
+from typing import Any, Dict, Iterable, List, Optional, Tuple, Type
+
+from django.db import models
+from django.http import HttpRequest, HttpResponse
+from django.contrib import messages
+from django import forms
+from django.shortcuts import get_object_or_404
+from django.utils.text import format_lazy
+from django.utils.translation import ugettext_lazy as _, ngettext
+
+from course.models import CourseModule, UserTag
+from course.viewbase import CourseInstanceMixin, CourseInstanceBaseView
+from deviations.models import SubmissionRuleDeviation
+from lib.viewbase import BaseFormView, BaseRedirectView
+from authorization.permissions import ACCESS
+from exercise.models import BaseExercise
+from userprofile.models import UserProfile
+
+
+class ListDeviationsView(CourseInstanceBaseView):
+    access_mode = ACCESS.TEACHER
+    deviation_model: Type[SubmissionRuleDeviation]
+
+    def get_common_objects(self) -> None:
+        super().get_common_objects()
+        all_deviations = self.deviation_model.objects.filter(
+            exercise__course_module__course_instance=self.instance
+        )
+        self.deviation_groups = get_deviation_groups(all_deviations)
+        self.note("deviation_groups")
+
+
+class AddDeviationsView(CourseInstanceMixin, BaseFormView):
+    access_mode = ACCESS.TEACHER
+    deviation_model: Type[SubmissionRuleDeviation]
+    session_key: str
+
+    def get_form_kwargs(self) -> Dict[str, Any]:
+        kwargs = super().get_form_kwargs()
+        kwargs["instance"] = self.instance
+        return kwargs
+
+    def form_valid(self, form: forms.BaseForm) -> HttpResponse:
+        exercises = get_exercises(form.cleaned_data)
+        submitters = get_submitters(form.cleaned_data)
+        existing_deviations = self.deviation_model.objects.filter(
+            exercise__in=exercises,
+            submitter__in=submitters,
+        )
+
+        if existing_deviations:
+            # Some deviations already existed. Use OverrideDeviationsView to
+            # confirm which ones the user wants to override. Store the form
+            # values in the current session, so they can be used afterwards.
+            self.success_url = self.deviation_model.get_override_url(self.instance)
+            self.request.session[self.session_key] = self.serialize_session_data(form.cleaned_data)
```

Comment:
There were several options for what to store in the session cache:
- Input form data (this is what I did)
- "Expanded" input form data, i.e. module IDs are expanded to exercise IDs and user tag IDs are expanded to user IDs
- The new deviation objects as JSON
- List of overlapping deviation IDs and new values from the form data (this is what the old PR did)

I ended up going with the first option because it keeps the session cache relatively small and the code mostly straightforward. I didn't like what the old PR did, because it assumed that the database table would stay unchanged between submitting the "add" and "override" forms. My solution is more "stateless".